core/storage/question/gae_models.py

```diff
@@ -283,62 +282,68 @@ def get_question_skill_links_based_on_mastery_equidistributed_by_skill(
-                if degree_of_mastery is None:
-                    degree_of_mastery = 0.0
-
             query = cls.query(cls.skill_id == skill_id)
 
-            # Fetch QuestionSkillLinkModels with difficulty larger than
-            # mastery and sort them by increasing difficulty.
-            harder_questions_query = query.filter(
-                cls.skill_difficulty >= degree_of_mastery)
-            harder_questions_query = harder_questions_query.order(
-                cls.skill_difficulty)
+            equal_questions_query = query.filter(
+                cls.skill_difficulty == difficulty_requested)
+            # Get as many requested difficulty questions as possible.
             new_question_skill_link_models = (
-                harder_questions_query.fetch(question_count_per_skill))
-
-            # Fetch QuestionSkillLinkModels with difficulty smaller than
-            # mastery and sort them by decreasing difficulty.
-            easier_questions_query = query.filter(
-                cls.skill_difficulty < degree_of_mastery)
-            easier_questions_query = easier_questions_query.order(
-                -cls.skill_difficulty)
-            new_question_skill_link_models.extend(
-                easier_questions_query.fetch(question_count_per_skill))
-
-            # Deduplicate if the same question is linked to multiple skills.
-            # Note: This deduplication might make the number of returning
-            # questions too low, but since we fetched double size of models
-            # above, the chance should be fairly small.
+                equal_questions_query.fetch(question_count_per_skill * 2))
             for model in new_question_skill_link_models:
                 if model.question_id in existing_question_ids:
                     new_question_skill_link_models.remove(model)
 
-            # Sort QuestionSkillLinkModels by the difference between their
-            # difficulty and mastery, and get the top
-            # question_count_per_skill number of them.
-            new_question_skill_link_models = sorted(
-                new_question_skill_link_models,
-                key=lambda model: abs(
-                    model.skill_difficulty - degree_of_mastery) # pylint: disable=cell-var-from-loop
-            )[:question_count_per_skill]
+            if len(new_question_skill_link_models) < question_count_per_skill:
+                # Fetch QuestionSkillLinkModels with difficulty smaller than
+                # requested difficulty and sort them by decreasing difficulty.
+                easier_questions_query = query.filter(
+                    cls.skill_difficulty < difficulty_requested)
+                easier_questions_query = easier_questions_query.order(
+                    -cls.skill_difficulty)
+                easier_question_skill_link_models = (
+                    easier_questions_query.fetch(question_count_per_skill))
+                for model in easier_question_skill_link_models:
```

Comment:
One thought about data structures: would it be more efficient to instead have a dict which maps question ID to the model? Then checking membership would be faster. You can convert it to a list at the end.

Reply:
Done.